### PR TITLE
Fix memory leak in ncdump

### PR DESCRIPTION
**Pull request by derobins**
ncdump is flagged by the gcc address sanitizer as having a small memory leak due to memory not being cleaned up before exiting.

It's been fixed to quiet the sanitizer output, but the ncdump program immediately exits after the ideal free location, so this isn't a problem for users.

Also quiets misc ncdump.c warnings.

### CODE REVIEW

File: mfhdf/ncdump/ncdump.c [hunk 1]
```diff
@@ -17,9 +17,9 @@
 
 static const char *type_name(nc_type type);
 
-char        *progname;
-struct ncdim dims[H4_MAX_NC_DIMS];  /* dimensions */
-long         vdims[H4_MAX_NC_DIMS]; /* dimension sizes for a single variable */
+char         *progname;
+struct ncdim *dims  = NULL; /* dimensions */
+long         *vdims = NULL; /* dimension sizes for a single variable */
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:22]: Dynamic arrays fix frame size warnings

File: mfhdf/ncdump/ncdump.c [hunk 2]
```diff
@@ -141,14 +141,14 @@ pr_att_vals(nc_type type, int len, void *vals)
     char         *sp;
     unsigned char uc;
     char          gps[30]; /* for ascii of a float or double precision */
-    char         *f_fmt = "%#.8g";
-    char         *d_fmt = "%#.16g";
+    const char   *f_fmt = "%#.8g";
+    const char   *d_fmt = "%#.16g";
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:145]: Const warning fixes

File: mfhdf/ncdump/ncdump.c [hunk 3]
```diff
@@ -161,7 +161,7 @@ pr_att_vals(nc_type type, int len, void *vals)
             while (*sp-- == '\0' && len > 0)
                 len--;
             for (iel = 0; iel < len; iel++)
-                switch (uc = *gp.cp++ & 0377) {
+                switch (uc = (unsigned char)(*gp.cp++ & 0377)) {
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:164]: Warning fixes

File: mfhdf/ncdump/ncdump.c [hunk 4]
```diff
@@ -388,7 +388,7 @@ do_ncdump(char *path, struct fspec *specp)
 
             printf("\t\t%s:%s = ", fixed_var, fixed_att);
             (void)ncattinq(ncid, varid, att.name, &att.type, &att.len);
-            att.val = (void *)malloc((unsigned)att.len * nctypelen(att.type));
+            att.val = (void *)malloc((size_t)(att.len * nctypelen(att.type)));
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:391]: size_t clears allocator warnings

File: mfhdf/ncdump/ncdump.c [hunk 5]
```diff
@@ -525,6 +525,8 @@ do_ncdump(char *path, struct fspec *specp)
         }
     }
 
+    free(vlist);
+
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:529]: This free call fixes the ncdump memory leak

File: mfhdf/ncdump/ncdump.c [hunk 6]
```diff
@@ -542,7 +544,7 @@ make_lvars(char *arg, struct fspec *fspecp)
         if (*cp == ',')
             nvars++;
 
-    fspecp->lvars = (char **)malloc(nvars * sizeof(char *));
+    fspecp->lvars = (char **)malloc((size_t)nvars * sizeof(char *));
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:547]: Fixes a signed allocation size warning

File: mfhdf/ncdump/ncdump.c [hunk 7]
```diff
@@ -574,8 +576,8 @@ set_sigdigs(char *arg)
     char *ptr2       = 0;
     long  flt_digits = 7;  /* default floating-point digits */
     long  dbl_digits = 15; /* default double-precision digits */
-    char  flt_fmt[6];
-    char  dbl_fmt[6];
+    char  flt_fmt[16];
+    char  dbl_fmt[16];
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:580]: Buffer was too small for the max int as chars

File: mfhdf/ncdump/ncdump.c [hunk 8]
```diff
@@ -620,6 +622,14 @@ main(int argc, char *argv[])
...
         usage();
 
+    dims  = (struct ncdim *)calloc(H4_MAX_NC_DIMS, sizeof(struct ncdim)); /* dimensions */
+    vdims = (long *)calloc(H4_MAX_NC_DIMS, sizeof(long)); /* dimension sizes for a single variable */
+
+    if ((dims == NULL) || (vdims == NULL)) {
+        error("Unable to allocate memory!\n");
+        exit(EXIT_FAILURE);
+    }
+
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:632]: Dynamic arrays

File: mfhdf/ncdump/ncdump.c [hunk 9]
```diff
@@ -669,7 +679,7 @@ main(int argc, char *argv[])
                 }
                 break;
             case 'l': /* maximum line length */
-                max_len = strtol(h4optarg, 0, 0);
+                max_len = (int)strtol(h4optarg, 0, 0);
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:682]: Warning cast

File: mfhdf/ncdump/ncdump.c [hunk 10]
```diff
@@ -698,5 +708,8 @@ main(int argc, char *argv[])
             do_ncdump(argv[i], &fspec);
     } while (++i < argc);
 
+    free(dims);
+    free(vdims);
+
```
Comment (derobins) [inline mfhdf/ncdump/ncdump.c:713]: Dynamic arrays

**Review by byrnHDF — APPROVED** [review]

**Pull request merged**